onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -297,35 +298,49 @@ def enketo(self, request, *args, **kwargs):
 
     def destroy(self, request, *args, **kwargs):
         instance_ids = request.data.get('instance_ids')
+        delete_all_submissions = request.data.get('delete_all', False)
         self.object = self.get_object()
 
         if isinstance(self.object, XForm):
-            if not instance_ids:
+            if not instance_ids and delete_all_submissions == 'False':
                 raise ParseError(_(u"Data id(s) not provided."))
             else:
-                instance_ids = [
-                    x for x in instance_ids.split(',') if x.isdigit()]
-
-                if not instance_ids:
-                    raise ParseError(_(u"Invalid data ids were provided."))
-
                 initial_count = self.object.submission_count()
-                queryset = Instance.objects.filter(
-                    id__in=instance_ids,
-                    xform=self.object,
-                    # do not update this timestamp when the record have
-                    # already been deleted.
-                    deleted_at__isnull=True
-                )
-                # loop through queryset
-                # then call delete_instance that calls .save()
-                # to allow emitting post_save signal
-                for instance in queryset_iterator(queryset):
-                    delete_instance(instance, request.user)
+                if delete_all_submissions == 'True':
+                    # Update timestamp only for active records
+                    self.object.instances.filter(
+                        deleted_at__isnull=True).update(
+                            deleted_at=timezone.now(),
+                            deleted_by=request.user)
+                else:
+                    instance_ids = [
+                        x for x in instance_ids.split(',') if x.isdigit()]
+                    if not instance_ids:
+                        raise ParseError(_(u"Invalid data ids were provided."))
+
+                    self.object.instances.filter(
+                        id__in=instance_ids,
+                        xform=self.object,
+                        # do not update this timestamp when the record have
+                        # already been deleted.
+                        deleted_at__isnull=True
+                    ).update(
+                        deleted_at=timezone.now(),
+                        deleted_by=request.user)
+
                 # updates the num_of_submissions for the form.
                 after_count = self.object.submission_count(force_update=True)
                 number_of_records_deleted = initial_count - after_count
 
+                # update the date modified field of the project
+                self.object.project.date_modified = timezone.now()
+                self.object.project.save(update_fields=['date_modified'])
+
+                # Track submissions made today
+                _update_submission_count_for_today(
```

Comment:
Why do we update this for all deletes ? _I'd expect that we'd only decrease the submission count of the day if a submission created on that day was deleted ?_

Reply:
_Precautionary measure If I would say_. To handle instances of when all submissions are deleted and a submission was created on that day.
The submission count for the day is a cached value, so I dont see a downside of having this here

Reply:
Their is a downside of using it here... It'll end up causing the submission count for the day to be inaccurate by 1 submission; _The function doesn't actually run any query to check when submissions were created... or how many submissions were deleted_